webapp/login/oauth_views.py

```diff
@@ -0,0 +1,47 @@
+import flask
+from webapp.decorators import login_required
+from webapp.api.github import GithubApi, GithubAuthException
+
+oauth = flask.Blueprint(
+    "oauth", __name__, template_folder="/templates", static_folder="/static",
+)
+
+github = GithubApi()
+
+
+@oauth.route("/github/auth", methods=["GET"])
+@login_required
+def github_auth():
+    if flask.request.args.get("back"):
+        flask.session["github_auth_redirect"] = flask.request.headers.get(
+            "Referer"
+        )
+
+    return github.authorize(scope="repo", state=flask.session["csrf_token"])
+
+
+@oauth.route("/github/auth/verify", methods=["GET"])
+@login_required
+def github_login_verify():
+    state = flask.request.args.get("state")
+
+    # Avoid CSRF attacks
+    if state != flask.session["csrf_token"]:
+        flask.flash("Invalid request", "negative")
+        return flask.redirect(flask.url_for("snapcraft.homepage"))
+
+    auth_code = flask.request.args.get("code")
+    auth_token = github.verify_auth(auth_code)
+    flask.session["github_auth_secret"] = auth_token
+
+    url = flask.session.pop(
+        "github_auth_redirect", flask.url_for("snapcraft.homepage")
+    )
+
+    return flask.redirect(url)
+
+
+@oauth.errorhandler(GithubAuthException)
```

Comment:
Again, this flow would be quite difficult for a developer to unpick. I think it would be much better to do this logic and redirection directly in the view where you expect the authentication to fail. This has the benefit that you won't end up catching and masking errors where you didn't expect with a redirect. If something happens that you didn't expect, the best possible thing that can happen is that your application breaks with a useful error message.